### PR TITLE
core: separate out decorator for anonymous rewrite pattern

### PR DESCRIPTION
**Pull request by superlopuh**
I never liked the overload based on number of parameters for the op type thing, this PR separates out the two flows. I think that this makes things a bit less confusing, there's one annotation for the method rewrite, and another annotation that builds a rewrite based on a standalone function.

I'm doing a bit of work in the rewrites, and having the two be separate also makes it easier for me to experiment, but that's not really the primary goal of the PR.

### CODE REVIEW

**Comment by math-fehr** [comment]
Actually, I have an alternative proposition. Should we just directly remove the anonymous rewrites? No one really use them, so let's remove them entirely? It would make the API more simple.

**Comment by math-fehr** [comment]
Yeah let's do this!
This was again one of my premature API optimizations that just wasn't useful!

**Comment by superlopuh** [comment]
Turns out I was the only one using it :)

File: xdsl/interpreters/experimental/pdl.py [hunk 1]
```diff
@@ -171,6 +173,35 @@ def match_operation(
         return True
 
 
+@dataclass
+class InterpreterRewrite(RewritePattern):
```
[the remark this answers is not in the document]
Comment (math-fehr) [reply xdsl/interpreters/experimental/pdl.py:177]: Yeah, both are fine for me here!

**Review by webmiche — APPROVED** [review]
Nice, you are doing the lords work (or whatever that was) 😉

**Pull request merged**